### ISSUE TITLE
project

### ISSUE THREAD

**Issue by nerkarkhushi16-beep** [nerkarkhushi16-beep/Srushti#1]
# High-Grade Concrete Compressive Strength Prediction Prototype
# Project Phase: Kickoff & Backend Initialization

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

print("=" * 65)
print("  HIGH-GRADE CONCRETE STRENGTH PREDICTION - ML PROTOTYPE ENGINE")
print("=" * 65)

# 1. Generate synthetic high-grade concrete benchmark data
print("\n[Phase 1] Initializing Data Pipeline & Feature Vectors...")

np.random.seed(42)
num_samples = 200

data = {
    "Cement_kg_m3": np.random.uniform(350, 550, num_samples),
    "Blast_Furnace_Slag_kg_m3": np.random.uniform(0, 150, num_samples),
    "Fly_Ash_kg_m3": np.random.uniform(0, 120, num_samples),
    "Water_kg_m3": np.random.uniform(140, 180, num_samples),
    "Superplasticizer_kg_m3": np.random.uniform(5, 15, num_samples),
    "Coarse_Aggregate_kg_m3": np.random.uniform(800, 1100, num_samples),
    "Fine_Aggregate_kg_m3": np.random.uniform(600, 850, num_samples),
    "Age_Days": np.random.choice([3, 7, 14, 28, 56, 90], num_samples),
}

df = pd.DataFrame(data)

# Simulate compressive strength using a synthetic relationship
df["Compressive_Strength_MPa"] = (
    (df["Cement_kg_m3"] * 0.12)
    + (df["Blast_Furnace_Slag_kg_m3"] * 0.05)
    + (df["Fly_Ash_kg_m3"] * 0.04)
    - (df["Water_kg_m3"] * 0.25)
    + (df["Superplasticizer_kg_m3"] * 1.5)
    + (df["Age_Days"] * 0.3)
    + np.random.normal(15, 3, num_samples)
)

print(f"-> Successfully simulated {num_samples} experimental mix entries.")
print(
    f"-> Strength range: "
    f"{df['Compressive_Strength_MPa'].min():.1f} to "
    f"{df['Compressive_Strength_MPa'].max():.1f} MPa."
)

# 2. Feature extraction and splitting
X = df.drop(columns=["Compressive_Strength_MPa"])
y = df["Compressive_Strength_MPa"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
)

# 3. Normalize features for linear regression
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# 4. Train and compare models
print("\n[Phase 2] Training ML Algorithms...")

linear_model = LinearRegression()
linear_model.fit(X_train_scaled, y_train)
linear_predictions = linear_model.predict(X_test_scaled)

random_forest_model = RandomForestRegressor(
    n_estimators=100,
    random_state=42,
    n_jobs=-1,
)
random_forest_model.fit(X_train, y_train)
random_forest_predictions = random_forest_model.predict(X_test)

print("-> Models trained successfully.")

# 5. Evaluate models
models_metrics = {
    "Linear Regression": {
        "R2": r2_score(y_test, linear_predictions),
        "RMSE": np.sqrt(mean_squared_error(y_test, linear_predictions)),
    },
    "Random Forest": {
        "R2": r2_score(y_test, random_forest_predictions),
        "RMSE": np.sqrt(mean_squared_error(y_test, random_forest_predictions)),
    },
}

print("\n" + "=" * 65)
print("                    MODEL COMPARISON")
print("=" * 65)
print(f"{'Model':<25} | {'R2 Score':<12} | {'RMSE (MPa)':<12}")
print("-" * 65)

for model_name, metrics in models_metrics.items():
    print(
        f"{model_name:<25} | "
        f"{metrics['R2']:<12.4f} | "
        f"{metrics['RMSE']:<12.2f}"
    )

print("=" * 65)

# 6. Feature importance
print("\n[Phase 3] Feature Importance from Random Forest:")

importance_df = (
    pd.DataFrame(
        {
            "Feature": X.columns,
            "Importance": random_forest_model.feature_importances_,
        }
    )
    .sort_values(by="Importance", ascending=False)
)

for _, row in importance_df.iterrows():
    print(
        f" - {row['Feature']:<30}: "
        f"{row['Importance'] * 100:>6.2f}%"
    )

print("\n" + "=" * 65)
print(" STATUS: Backend framework ready. Infrastructure operational.")
print("=" * 65)
